pkg/admissioncontroller/webhooks/validate_kubeconfig_secrets.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhooks
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/gardener/gardener/pkg/admissioncontroller/metrics"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+const (
+	kubeconfigValidatorName = "kubeconfig_validator"
+
+	// reasonInvalidKubeconfig is a StatusReason that will be sent as part of the webhook's response if the secret
+	// contains an invalid kubeconfig
+	reasonInvalidKubeconfig = metav1.StatusReason("InvalidKubeconfig")
+)
+
+var secretGVK = metav1.GroupVersionKind{Group: "", Kind: "Secret", Version: "v1"}
+
+var _ admission.Handler = &KubeconfigSecretValidator{}
+
+// KubeconfigSecretValidator is a webhook handler validating CREATE and UPDATE requests on secrets. It checks, if the
+// secrets contains a kubeconfig and denies kubeconfigs with invalid fields (e.g. tokenFile or exec).
+type KubeconfigSecretValidator struct {
+	logger  logr.Logger
+	decoder *admission.Decoder
+}
+
+// InjectLogger injects a logger into the handler.
+func (h *KubeconfigSecretValidator) InjectLogger(l logr.Logger) error {
+	h.logger = l.WithName(kubeconfigValidatorName)
+	return nil
+}
+
+// InjectDecoder injects a decoder capable of decoding objects included in admission requests.
+func (h *KubeconfigSecretValidator) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
```

Comment:
Isn't this injection stuff deprecated? Should we rather not use it to prevent follow-up efforts? (same below)

Reply:
I started like this, before I got aware of the deprecation. I would leave it like this for now, as there are many other places we will have to adapt as well and there is still a bit of time left until deprecation.
Would that be fine for you?

Reply:
OK, sure.